Refactor hull

agoss94 opened the pull request:
Refactoring of commons-geometry-hull module. The ConvexHull interface has been relocated to commons-geometry-core and ConvexHull2D is now part of commons-geometry-euclidean with a builder that combines the logic of the Akl-Toussaint heuristic and the Monotone-Chain to build a convex hull.

Line 291-295 in ConvexHull2D has to be added to the original code in order to allow for collinear points to be considered in all cases. 

All necessary test have been moved to the euclidean module as well.

aherbert commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,469 @@
+         */
+        public Builder append(Vector2D point) {
+
+            //Checks if the given point supersedes one of the corners.
```
Whitespace after `//`. Same again on the next comment below.

agoss94 replied:
Fixed

aherbert commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,469 @@
+         * @param point a given point.
+         */
+        private void checkCorners(Vector2D point) {
+            if (minX == null || point.getX() < minX.getX()) {
```
Here we only require one null check. If minX is null then the other corners are also null and all can be set to the point:
```Java
if (minX == null) {
    minX = maxX = minY = maxY = point;
    candidates.add(point);
    return;
}
```
IIUC this condition will also be met if `candidate.isEmpty()`. However using the null check does indicate that for the rest of the method the corners are not null.

agoss94 replied:
The observation is correct I changed the code. Although the method becomes a bit longer this way we can save 3 of the four null checks as suggested.

aherbert commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,469 @@
+     */
+    public static final class Builder {
+
+        /** Corner of triangle with minimal x coordinate. */
```
Q. `triangle` to `quadrilateral`?

agoss94 replied:
Yes, that should obviously mean quadrilateral.

aherbert commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,469 @@
+        private boolean insideQuadrilateral(final Vector2D point,
+                                                   final List<? extends Vector2D> quadrilateralPoints) {
+
+            Vector2D v1 = quadrilateralPoints.get(0);
```
Since this walks a circle around the quadrilateral we can start anywhere. IIUC this can be changed to:
```java
final int size = quadrilateralPoints.size();
Vector2D v1 = quadrilateralPoints.get(size - 1);
Vector2D v2 = quadrilateralPoints.get(0);

// ...

for (int i = 1; i < size; i++) {
    v1 = v2;
    v2 = quadrilateralPoints.get(i);
```

agoss94 replied:
Yes, the change could be made here. More importantly it showed that I missed an edge case when collinear points should be included. I fixed that as well.

aherbert commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,469 @@
+            // concatenate the lower and upper hulls
+            // the last point of each list is omitted as it is repeated at the beginning of the other list
+            final List<Vector2D> hullVertices = new ArrayList<>(lowerHull.size() + upperHull.size() - 2);
+            for (int idx = 0; idx < lowerHull.size() - 1; idx++) {
```
Can these loops be changed to omit the first point of each half:
```java
for (int idx = 1; idx < lowerHull.size(); idx++) {
```

agoss94 replied:
I am going to check that tomorrow .

agoss94 replied:
Changing which point to omit was no problem.

aherbert reviewed (commented):
Thanks for the update. I've suggested a few optimisation changes (untested). Please review and see if they make sense.

One issue is a loss of the ConvexHull2D commit history. The file does have 5 commits which all have a Geometry Jira ticket. Keeping the history allows the changes to be followed back to the original version. The test classes have been moved. It is only this file that is deleted and added. You should be able to: checkout the deleted file from a commit before it was deleted; use `git mv` to move the new file to avoid a clash; and then again to move the old file to the new location; replace the contents with what you have; and delete the temp new file.

agoss94 commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
Old file has been retrieved and moved to the new location as suggested.

agoss94 commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,469 @@
+         * @return {@code true} if the point is inside the quadrilateral, {@code false} otherwise
+         */
+        private boolean insideQuadrilateral(final Vector2D point,
+                                                   final List<? extends Vector2D> quadrilateralPoints) {
```
Fixed

agoss94 commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,469 @@
+         */
+        private final boolean includeCollinearPoints;
+
+        /**Return a {@link Builder} instance configured with the given precision
```
Fixed

agoss94 commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,469 @@
+            while (it.hasNext()) {
+                p3 = it.next();
+
+                v1 = p1.vectorTo(p2);
```
I am going to check that tomorrow .

agoss94 replied:
Did you mean to precompute v1? I did that an then made the assigment at the end of the loop which worked without a problem.

aherbert replied:
I did mean precompute v2, then inside the loop shift v2 to v1 at the start and compute v2 again.

However you have precomputed v1, and shifted v2 to v1 at the end of the loop which is effectively the same (i.e. do not duplicate computing Vectors). I prefer your code as all the shifting of variables is in one place.

Note that you did not drop `p1 = p2` inside the loop (where p1 is now redundant).

I'll try and review fully in the next few days.

agoss94 replied:
Ah I see. I did miss the assignment. I deleted the line and renamed the variables to better fit the now altered code. Thanks for pointing out.

agoss94 commented:
All review marks have been completed. Please review all subsequent changes made.

darkma773r commented:
Hello! I plan on reviewing this soon, although I might not have time until this weekend. Thanks for the hard work here!

aherbert commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,475 @@
+                return this;
+            }
+
+            final List<Vector2D> quadrilateral = buildQuadrilateral(minY, maxX, maxY, minX);
```
It may be useful to cache the quadrilateral. Modification of the `checkCorners` method can be done to remove the cached quadrilateral if the point supersedes one of the corners.

agoss94 replied:
Or maybe get rid of the list at all. The way I see it. The purpose of the list is to count how many of the points are truly distinct and to go around all sides in an ordered manner. Since we only have 4 corners at all and only proceed if there are  at least 3 sides it would be possible to get rid of the list and just to count and then replace the loop with a method that checks either all three or four sides.

agoss94 replied:
It isn't really a new approach, but maybe more the fact, that the loop has only 3-4 iterations which could be straight forward to code. Since we only check the sides [minY, maxX], [maxX, maxY], [maxY, minX], [minX, minY]  for any given point, whereas we skip any side if the two points align, which can happen only once. I guess one could implement this logic entirely without a loop. 
I guess the simplest way is still to just cache the list.

aherbert commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,475 @@
+
+            while (it.hasNext()) {
+                p2 = it.next();
+
```
Remove this empty line.

aherbert reviewed (commented):
This is looking good. I note that it is possible to cache the quadrilateral. For larger clouds of points this may be useful when the quadrilateral is unchanged by many successive points.

Let's wait for a second opinion from Matt.

darkma773r commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,475 @@
+        final StringBuilder sb = new StringBuilder();
+        sb.append(getClass().getSimpleName())
+            .append("[vertices= ")
+            .append(getVertices())
```
That's probably a good idea. If we do this here, we should update (in another ticket) other classes that currently produce arbitrarily large `toString` representations, like `LinePath`. They should be consistent in the number of vertices printed before truncation.

agoss94 replied:
I would suggest to refrain from the change for now and to open a new ticket for this one. Gilles does have a different opinion on this topic (https://issues.apache.org/jira/projects/GEOMETRY/issues/GEOMETRY-110?filter=allopenissues) and even if we implement the change all the other classes still need an update. Better to change all the classes in one ticket.

darkma773r replied:
That sounds good to me.

darkma773r commented:
This looks great! Thank you, @agoss94! I think the only things to address are the two points made above (toString and caching the quadrilateral) and the code coverage. There are several if statement branches in `ConvexHull2D` that are not hit by the tests, which makes me think that either the tests are not sufficiently hitting all of the edge cases or those statements are not needed.

agoss94 commented:
@darkma773r I added a few test and removed some if branches which seemed to be really unnecessary. Most branches were not hit because those checked for identical points, which is superfluous as the PointsSet already filtered thoses out. 
Other if branches were not needed because of the comparator which orders the points before constructing the lower and upper hull. This allows us to make assumtions on the calculated distance to previously added points. 
You may double check the logic in deleting these checks.

darkma773r commented:
This looks good to me! @aherbert, any objections to merging this in (after a rebase, it looks like)?

aherbert commented:
I think the cache of the quadrilateral can be improved. It only has to be rebuilt if the 4 points for the extrema have changed. Here the builder is recreating it each time. IIUC if the corners change then you can rebuild the quadrilateral and add to the candidates. The rest of the work in `append` is not relevant and you can simply return. This can be done by returning a `boolean` from `checkCorners`, e.g.

```java
public Builder append(Vector2D point) {
    // Checks if the given point supersedes one of the corners.
    if (checkCorners(point)) {
        buildQuadrilateral(minY, maxX, maxY, minX);
        return this;
    }

    // if the quadrilateral is not well formed, e.g. only 2 points, do not attempt to reduce
    if (quadrilateral.size() < 3) {
        // Point cannot yet be dismissed.
        candidates.add(point);
        return this;
    }

    // check all points if they are within the quadrilateral
    // in which case they can not be part of the convex hull
    if (!insideQuadrilateral(point, quadrilateral)) {
        candidates.add(point);
    }

    return this;
}
```

agoss94 commented:
>
Proposition has been implemented. Thanks for the detailed review :) I learned a lot. 
The branch has also been rebased.

aherbert commented on commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java:
```diff
@@ -0,0 +1,461 @@
+         *         {@code false} otherwise.
+         */
+        boolean checkCorners(Vector2D point) {
+            boolean hasBeenModified = false;
```
Note: You do not require this flag until after your `min == null` check

aherbert commented:
@darkma773r I have no further suggestion for this. If you are OK with the new API then I suggest merging.
